### ISSUE TITLE
log(diag_part(x)) vs diag_part(log(x)) difference with Cholesky gradients

### ISSUE THREAD

**Issue by sethnabarro** [tensorflow/tensorflow#37577]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Mac OS Mojave 10.14.5
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: -
- TensorFlow installed from (source or
binary): binary (by pip)
- TensorFlow version (use command below): 2.1.0
- Python version: 3.6.6
- Bazel
version (if compiling from source): -
- GCC/Compiler version (if compiling from
source): -
- CUDA/cuDNN version: - (running on CPU)
- GPU model and memory: - (running on CPU)

**Describe the current behavior**
Gradients of a variable, whose loss depends on the logarithm of the diagonal of its Cholesky decomposition (`tf.linalg.cholesky()`), can fail depending on the ordering of the `tf.math.log()` and `tf.linalg.diag_part()` operations. `tf.linalg.diag_part(tf.math.log(L))` results in NaN gradients, `tf.math.log(tf.linalg.diag_part(L))` runs as expected (where `L` is the Cholesky decomposition). 

**Describe the expected behavior**
`tf.math.log(tf.linalg.diag_part(L))` and `tf.linalg.diag_part(tf.math.log(L))` should both give rise to the same (correct) gradients. the order of these ops shouldn't matter

**Standalone code to reproduce the issue** 
```
import numpy as np
import tensorflow as tf


LR = 0.01
X = tf.random.normal([100, 10], 0., 1.)
LOG2PI = 1.8378770664093453
WITH_BUG = True
N_ITERS = 100


def gaussian_pdf(samples, mean, covariance):
    n = tf.cast(tf.shape(samples)[1], tf.float32)
    L = tf.linalg.cholesky(covariance)
    alpha = tf.linalg.cholesky_solve(L, tf.transpose(samples - mean))
    data_fit = -0.5 * tf.reduce_sum(tf.transpose(alpha) * (samples - mean), -1)
    regulariser_bug = -0.5 * tf.reduce_sum(tf.linalg.diag_part(tf.math.log(L)), axis=-1)
    regulariser_fine = -0.5 * tf.reduce_sum(tf.math.log(tf.linalg.diag_part(L)), axis=-1)
    regulariser = regulariser_bug if WITH_BUG else regulariser_fine
    normaliser = -n * 0.5 * LOG2PI
    return data_fit + regulariser + normaliser


def loss(x, mu, sigma):
    ll = gaussian_pdf(x, mu, sigma)
    return -tf.reduce_mean(ll)


cov_np = np.random.normal(size=(10, 10), scale=0.001) + np.eye(10) * 2.
cov_var = tf.Variable(cov_np, dtype=tf.float32)
mean_var = tf.Variable([3.] * 10, dtype=tf.float32)


def train(x):
    with tf.GradientTape() as t:
        nll = loss(x, mean_var, cov_var)
    dm, dsig = t.gradient(nll, [mean_var, cov_var])
    dm_avg, dcov_avg = tf.reduce_mean(dm).numpy(), tf.reduce_mean(dsig).numpy()
    print(f'Avg grads: mu: {dm_avg}, cov: {dcov_avg}')
    mean_var.assign_sub(LR * dm)
    cov_var.assign_sub(LR * dsig)


for _ in range(N_ITERS):
    print(f'Negative log-likelihood: {loss(X, mean_var, cov_var).numpy()}')
    train(X)
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF2.1](https://colab.research.google.com/gist/amahendrakar/a5a82eabde0cfb98d3c0ff122dbb915c/37577.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/1c85bd9b1be3006a1dff270d9f1f4254/37577-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by saxenasaurabh** [comment]
Thanks @sethnabarro for reporting this. This seems to be an artifact of how the gradient of `tf.diag_part` is [implemented](https://github.com/tensorflow/tensorflow/blob/6bc4800a402f54703e0a6dd0a444769d61910980/tensorflow/python/ops/array_grad.py#L387). We fill the non-diagonal values with zeros and if those entries in `L` are zeros as well you would get a `0/0`. I don't see an easy way to fix this for now (unless we add some fused APIs which does not seem combinatorially feasible) but I will discuss possible solutions with other folks.

**Comment by sethnabarro** [comment]
Hi @saxenasaurabh, thanks for your response. Ok, I understand the issue. I will also have a think about possible solutions.

**Comment by kumariko** [comment]
Was able to reproduce the issue with TF2.7.0 . Please find the gist [here](https://colab.research.google.com/gist/kumariko/a09c2dc30d8dbdeeefda2a7616e53ebc/37577.ipynb#scrollTo=wUIpG5oBCpJp). Thanks!

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
